<a href="https://colab.research.google.com/github/wiradear/Prinsiple-of-DataScience/blob/main/Assingment_02_Data_Cleansing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## CS245 : Principle of Data Science
## <u>Assignment#02</u>: Data Cleansing
### (100 คะแนน)

We will use the Hotel Booking Demand dataset from the Kaggle.
The data file ``02_hotel_booking.csv`` has been attached to this assignment.
This data set contains booking information for a *city hotel* and a *resort hotel* and includes information such as when the booking was made, lead time, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. Descriptions for every column in the data set has made available with this assignment as well.

####1. (1 คะแนน) Import Required Packages

Bring in all the required packages.

In [ ]:
#Q.1.
import pandas as pd

####2. (1 คะแนน) Input Data

Read the data set into our working area as a ``dataframe`` named *booking*.

In [ ]:
#Q.2.
from google.colab import drive
drive.mount('/content/drive')
booking = pd.read_csv('/content/drive/MyDrive/CS245/02_hotel_bookings.csv')
booking

Mounted at /content/drive


hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

###3. (10 คะแนน) Summary Information About Data Set

####3.1. (3 คะแนน) Display Summary Information
Use Python instructions to display summary information about the data set.
Basic information such as the number of rows and columns, detailed information about each column and descriptive statistical details of the data set must be displayed. Also, the first 10 rows in the data set must be shown.

In [ ]:
# 3.1
# Basic information such as the number of rows and columns
booking.head(10)


hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   
5  Resort Hotel            0         14               2015               July   
6  Resort Hotel            0          0               2015               July   
7  Resort Hotel            0          9               2015               July   
8  Resort Hotel            1         85               2015               July   
9  Resort Hotel            1         75               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   
5                        27                          1   
6                        27                          1   
7                        27                          1   
8                        27                          1   
9                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   
5                        0                     2       2  ...    No Deposit   
6                        0                     2       2  ...    No Deposit   
7                        0                     2       2  ...    No Deposit   
8                        0                     3       2  ...    No Deposit   
9                        0                     3       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type    adr  \
0    NaN     NaN                    0     Transient    0.0   
1    NaN     NaN                    0     Transient    0.0   
2    NaN     NaN                    0     Transient   75.0   
3  304.0     NaN                    0     Transient   75.0   
4  240.0     NaN                    0     Transient   98.0   
5  240.0     NaN                    0     Transient   98.0   
6    NaN     NaN                    0     Transient  107.0   
7  303.0     NaN                    0     Transient  103.0   
8  240.0     NaN                    0     Transient   82.0   
9   15.0     NaN                    0     Transient  105.5   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   
5                            0                          1           Check-Out   
6                            0                          0           Check-Out   
7                            0                          1           Check-Out   
8                            0                          1            Canceled   
9                            0                          0            Canceled   

  reservation_status_date  
0     

In [ ]:
# 3.1
#detailed information about each column
booking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [ ]:
# 3.1
#descriptive statistical details of the data set must be displayed
booking.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000                   0.000000  
50%        94.57500

####3.2 (4 คะแนน) ชุดข้อมูลนี้มีข้อมูลที่ขาดหายไป (missing values) หรือ NaN หรือไม่? ถ้ามี คอลัมน์ใดบ้างที่มีข้อมูลขาดไป จำนวนเท่าใด  อภิปรายเหตุผลประกอบคำตอบของนักศึกษา

(นักศึกษาตอบคำถามข้อ 3.2 ใน section นี้ -- ตอบเป็นภาษาไทยได้)

ตอบ มี missing value จากการดูข้อมูลที่ได้จาก booking.info() ในส่วนของ  Non-Null Count จะเห็นได้ว่า มี missing value ในส่วนของ

*   column children มี 119386 non-null จากทั้งหมด 119390 = ขาดไป 4
*   column  country มี 118902 non-null จากทั้งหมด 119390 = ขาดไป 488
*   column agent มี 103050 non-null จากทั้งหมด 119390 = ขาดไป 16340
*   column company มี 6797 non-null  จากทั้งหมด 119390 = ขาดไป 112593

หรือดูได้จากการได้ฟังก์ชัน isnull() และ sum() เพื่อดูจำนวนของ missing value/Nan









####3.3 (3 คะแนน) เมทอดที่ให้ค่าสถิติของชุดข้อมูล (Data set) ข้างต้น คำนวณค่าสถิติทุกคอลัมน์ข้อมูลหรือไม่ อภิปรายเหตุผลประกอบด้วย

(นักศึกษาตอบคำถามข้อ 3.3 ใน section นี้ -- ตอบเป็นภาษาไทยได้)

ตอบ ถ้าเป็นเมทอด .describe() จะไม่ได้คำนวณค่าสถิติทุกคอลัมน์ข้อมูล กล่าวคือ จะประมาณผลเฉพาะคอลัมน์ที่เป็นตัวเลขและไม่คำนวณคอลัมน์ที่มีค่าเป็น NaN



---


###4. (43 คะแนน) Data Cleansing

####4.1. (38 คะแนน) Handling Missing Values

Use appropriate Python codes to accomplish the work listed in Q.4.1.1 - Q.4.1.7


##### 4.1.1. (3 คะแนน) Create a Boolean index (of the same size as the original dataframe) indicating NaN fields in the data set

[use the following code area for answering this question]

In [ ]:
# Q.4.1.1.
print(booking.isnull()) #use isnull() to check, if result = True = มี missing value

        hotel  is_canceled  lead_time  arrival_date_year  arrival_date_month  \
0       False        False      False              False               False   
1       False        False      False              False               False   
2       False        False      False              False               False   
3       False        False      False              False               False   
4       False        False      False              False               False   
...       ...          ...        ...                ...                 ...   
119385  False        False      False              False               False   
119386  False        False      False              False               False   
119387  False        False      False              False               False   
119388  False        False      False              False               False   
119389  False        False      False              False               False   

        arrival_date_week_number  arriv


##### 4.1.2. (5 คะแนน) the number of NaN values in every column sorted descendingly (i.e. show the information corresponding to the columns with highest number of missing values first)

[use the following code area for answering this question]

In [ ]:
# Q.4.1.2.
print(booking.isnull().sum().sort_values(ascending=False)) #sorted number of missing value in every column from highest number

company                           112593
agent                              16340
country                              488
children                               4
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
hotel                                  0
previous_cancellations                 0
days_in_waiting_list                   0
customer_type                          0
adr                                    0
required_car_parking_spaces            0
total_of_special_requests              0
reservation_status                     0
previous_bookings_not_canceled         0
is_repeated_guest                      0
is_canceled                            0
distribution_channel                   0
market_segment                         0
meal                                   0
babies                                 0
adults                                 0
stays_in_week_ni

##### 4.1.3. (5 คะแนน) Is there any row in the data set with all NaNs?  Design your Python script to answer this question. Students must provide comments along your code to explain your solution.
```
Expected Output:
Number of rows with all NaNs = ..........................
```

[use the following code area for answering this question]

In [ ]:
# Q.4.1.3.
print("Number of rows with all NaNs =",booking.isnull().all(axis=1).sum()) #ตรวจสอบทุกตัว(all())ใน row นั้นว่าข้อมูลในทุกคอลัมน์ของ row นั้นเป็น Nan หรือไม่และใช้ sum() บวกจำนวนของ row ที่ทุกcolumnเป็น NaN

Number of rows with all NaNs = 0


##### 4.1.4. (7 คะแนน) Fill missing values in every column (except the 'country' and 'children' column) with ``0``.
Show how you verify that all NaN values in the specified columns have been replaced.

Students must provide comments along your code to explain your solution.  

[use the following code area for answering this question]

In [ ]:
# Q.4.1.4.
for cols in booking.columns: #loop to check all columns in data set
  if cols!= 'country' and cols!='children': #condition to check columns except country 'and' children
    booking[cols].fillna(0, inplace=True) #if column!= country and !=children , fill Nan/missing value with zero.
booking[cols].isnull().any().sum()
booking.info() # print info() to verify that all NaN values in the specified columns have been replaced โดยดูที่จำนวน Non-Nulll count




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

##### 4.1.5. (6 คะแนน) Fill all missing values in column 'country' with the most-frequently occurred (``mode``) value of the column.

Show how you verify that all NaN values in the column 'country' have been replaced.

Students must provide comments along your code to explain your solution.

[use the following code area for answering this question]


In [ ]:
# Q.4.1.5.
import statistics  #import statistics to use mode()
booking['country']=booking['country'].fillna(value=statistics.mode(booking['country']))  #replace NaN/missing values by mode(the most-frequently occurred) of country
booking.info() # look at Non-Null count of country to check that all Nan values in column 'country have been replaced , so country = 119390 non-null = ถูกแทนค่าเรียบร้อยแล้ว

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

#####4.1.6. (6 คะแนน) Fill all missing values in column 'children' with the rounded ``mean`` value of the column.

In [ ]:
# Q.4.1.6.
booking['children']=booking['children'].fillna(value = booking['children'].mean()) #fill boking['children'].mean() replace missing value/Nan
booking.info() #check Non-Null count of column children ถ้า children = 119390 non-null = ถูกแทนค่าเรียบร้อยแล้ว


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119390 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

##### 4.1.7. (3 คะแนน) Change data types of columns as follows:
  * 'children' : เปลี่ยนเป็นจำนวนเต็ม (integer)
  * 'agent' : เปลี่ยนเป็นจำนวนเต็ม (integer)
  * 'company' : เปลี่ยนเป็นจำนวนเต็ม (integer)

Show how you verify that the data types have been changed as expected.

Students must provide comments along your code to explain your solution.

[use the following code area for answering this question]

In [ ]:
# Q.4.1.7.
booking = booking.astype({"children":'int', "agent":'int', "company":'int'}) #เปลี่ยนชนิดของคอลัมน์ children agent และ company ให้มี type เป็น int
booking[['agent', 'children', 'company']].sum() #พิมพ์ผลรวมของจำนวนในแต่ละคอลัมน์ ไม่มีทศนิยม คือเป็นเลขจำนวนเต็ม และมี dtype บอกว่าเป็น int

agent       8933753
children      12403
company     1286446
dtype: int64

##### 4.1.8. (3 คะแนน) Verify that afterwards no missing values in the data set.

Students must provide comments along your code to explain your solution.  

[use the following code area for answering this question]

In [ ]:
# Q.4.1.8.
print(booking.isnull().sum()) #ใช้ isnull() เช็คว่ามีค่าที่เป็น missing value อยู่หรือไม่ และใช้ sum() ในการบวกจำนวนของตัว missing value ในแต่ละ column และถ้าเป็น 0 คือไม่มี missing value


hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                         

####4.2. (5 คะแนน) Handling Duplicates

Use appropriate Python codes to accomplish the work listed in Q.4.2.1 - Q.4.2.2

#####4.2.1. (2.5 คะแนน) Check whether there is a duplicate in the given data set.

If any, please provide the number of duplicate rows in the given data set.

```
Expected Output:
Number of duplicate rows = ..........................
```

In [ ]:
#Q.4.2.1.
print('Number of dulicate rows = ', booking.duplicated().sum()) #print number of duplicated data



Number of dulicate rows =  32013


#####4.2.2. (2.5 คะแนน) Drop all duplicate rows found in the given data set.

Verify also that all duplicate rows have been removed.

Students must provide comments along your code to explain your solution.  

[use the following code area for answering this question]

In [ ]:
#Q.4.2.2.
booking.drop_duplicates(inplace=True) #drop ตัวที่ข้อมูลซ้ำกับคอลัมน์อื่นๆ
print('Number of duplicate row = ',booking.duplicated().sum()) #count the number of duplicate rows by counting True และผลลัพธ์จะเป็น 0 เมื่อไม่มีrowที่ซ้ำ

Number of duplicate row =  0




---



###5. (20 คะแนน) Data Validity

There are many rows in the data set that have zero guests including adults, children and babies. These type of rows does not make any sense.

Find booking rows with total guests equal to zero and remove all of them.

#####5.1.(5 คะแนน) Identify how many rows in the data set that have zero guests including adults, children and babies.

```
Expected Output:
Number of rows with total 0 guests = ..........................
```

In [ ]:
#Q.5.1.
guest = booking[['children', 'adults', 'babies']].sum(axis=1)==0
print("Number of rows with total 0 guests = ", guest.sum())



Number of rows with total 0 guests =  166


#####5.2.(10 คะแนน) Identify and print all indices of rows in the data set that have zero guests including adults, children and babies.

```
Expected Output:
Indices of .... rows with total 0 guests are as follows:
<เอาท์พุตส่วนนี้ แต่ละแถวให้พิมพ์ index ของแถวที่มี total guests เป็น 0>
2224 2409 3181 3684 3708 4127 9376 31765 32029 32827
34849 37459 40392 40922 41020 41056 42452 42591 42871 42934
... เอาท์พุตถูกละไว้ ไม่ได้แสดงท้้งหมด ...
114908 115029 115091 116251 116534 117087
```

In [ ]:
#Q.5.2.
print('Indices of',guest.sum(),' rows with total 0 guests are as follows:')
booking.index[booking[['children', 'adults', 'babies']].sum(axis=1)==0]


Indices of 166  rows with total 0 guests are as follows:


Int64Index([  2224,   2409,   3181,   3684,   3708,   4127,   9376,  31765,
             32029,  32827,
            ...
            112471, 112558, 113188, 114583, 114908, 115029, 115091, 116251,
            116534, 117087],
           dtype='int64', length=166)

#####5.3.(10 คะแนน) Drop all rows in the data set that have zero guests including adults, children and babies.

Verify also that all rows with total 0 guests have been removed.

Students must provide comments along your code to explain your solution.  

<u>Hint</u>: ใช้ index ในการระบุแถวข้อมูลที่ต้องการ drop

[use the following code area for answering this question]

In [ ]:
#Q.5.3.
booking.drop(booking.index[booking[['children', 'adults', 'babies']].sum(axis=1)==0]) #ถ้ามี index ที่ผลรวมใน row ใน column children, adults, babies เป็น 0 ให้ drop row นั้น

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit      0       0   
1                          0       2  ...    No Deposit      0       0   
2                          1       1  ...    No Deposit      0       0   
3                          1       1  ...    No Deposit    304       0   
4                          2       2  ...    No Deposit    240       0   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit    394       0   
119386                     5       3  ...    No Deposit      9       0   
119387                     5       2  ...    No Deposit      9       0   
119388                     5       2  ...    No Deposit     89       0   
119389                     7       2  ...    No Deposit      9       0   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

###6. (23 คะแนน) Data Validity (More -- Challenge)

There might be some rows in the data set that there are children or baby guests without adults. These type of rows does not make any sense.

Find booking rows with such condition and remove all of them.

#####6.1.(18 คะแนน) Identify how many rows in the data set that have children or babies guests, but zero adults.

```
Expected Output:
Number of rows with children or babies guests, but zero adults = .............
```

In [ ]:
#Q.6.1.
print("Number of rows with children or babies guests, but zero adults =", booking[(booking['adults']==0) & ((booking['children']!=0) | (booking['babies']!=0))].shape[0])

Number of rows with children or babies guests, but zero adults = 219


#####6.2.(5 คะแนน) Drop all rows in the data set that have children or babies guests, but zero adults.

Verify also that all rows with total 0 guests have been removed.

Students must provide comments along your code to explain your solution.  

In [ ]:
#Q.6.2.
booking=booking.drop(booking[(booking['adults']==0) & ((booking['children']!=0) | (booking['babies']!=0))].index, axis=0) #ดรอปโดยการใช้ index
booking.reset_index(drop=True)
booking[(booking['adults']==0) & ((booking['children']!=0) | (booking['babies']!=0))].shape[0] # เช็คว่าดรอปถูกต้องตามเงื่อนไข ถ้าผลลัธ์มีค่าเป็น 0 แสดงข้อมูลที่ไม่เป็นตามเงื่อนไขถูกลบแล้ว


0

###7. (2 คะแนน) Save to ``csv`` file

Save the dataframe resulting from all data cleansing steps above to a csv file.

Name the file ``hotel_booking_cleaned.csv``. Be sure that we do not save the indices to csv file.

This file will be used as an input file for your next homework.

In [ ]:
#Q.7.
booking.to_csv("hotel_booking_cleaned.csv")